### ISSUE TITLE
Add pattern recognition

### ISSUE THREAD

**Issue by 0xilly** [Runix-Minecraft/Azathoth#3]
Move Runix's pattern recognition to The library

**Comment by 0xilly** [comment]
Might have to write my own way of doing this without using Runix's. Runix's energy system is to ingrande to the pattern congregation.

**Comment by josiahseaman** [comment]
The only cross over between pattern recognition and the energy listing that I know of is getMostValuableBlock().  This method isn't used anywhere and should just be deleted.  Give it a try.  I think these can definitely be separated.

**Comment by 0xilly** [comment]
First thing: Your alive?

Second: I'll look into it again IIRC I couldn't get it to work without moving tiers/energy into library which is not going to happen the library has no need to for that information.

**Comment by josiahseaman** [comment]
Yep, I'm alive :-)  I've just been focusing on work more.  I've been playing GoD Factory and Star Conflict for fun.  But both of those are getting old.  So I'll be circling back around to minecraft again.  Runix is a long term commitment for me, so I'll always come back around to it.

We had a conversation about the energy link to pattern recognition a while ago.  There's just the one unused dependency.  They can be separated very easily.  What cannot be separated (nor should you want to) is that both use a unified HashMap<WorldXYZ, Block> input and output interface.  I settled on this interface data structure because it was constant time to check membership, which happens frequently, and I can iterate over the set of pairs without complicated loop structures.  There are a couple places that deviate slightly from the interface, but overall I'm still happy with that decision.

**Comment by 0xilly** [comment]
For the sake of speed of getting this done as 1 to 1 as possible but I do want to do some thing along the lines of `PatternRegistry.registerPattern(modid, new fooPattern);`  later for sorting purposes

**Comment by josiahseaman** [comment]
Using Block classes to represent target patterns actually got a lot better in 1.7 because now you can make made-up placeholder instances of RunixFakeBlock (sp?) when you're trying to match against some complex criteria.  Sure, the system could be expanded on, but start with the basics, ensure that Runix can still run against the library.  That will give us a structurally solid place to build off of.  Libraries that are built from proven code that was made to solve a real problem are always better than libraries made to solve theoretical problems.

**Comment by 0xilly** [comment]
By sorting I meant Logging(in retrospect that would have been better) I want to log what mod is registering which pattern in the pattern registry. The sorting of patterns in the registry will be in a first come first serve basis totally dependent on when FML loads the mod.

**Comment by 0xilly** [comment]
Ok looks like I'm able to execute a pattern but I'm able to execute by just clicking anywhere in the world.
![goddanmit](https://cloud.githubusercontent.com/assets/1331314/5434094/1d31b8a4-841c-11e4-8878-43dc948922cd.png)

**Comment by josiahseaman** [comment]
Clicking anywhere anywhere?  It's not pattern recognition unless it returns negative on non-matching areas.

I like your status message :D

**Comment by 0xilly** [comment]
Right now I'm glad that it's executing what's inside a pattern now that it does that I can try figure out where (I)it got fucked up :D
